[io.opentelemetry] Add OTLP metrics and traces (#21219)
openhab/openhab-addons: bundles/org.openhab.io.opentelemetry/README.md

---
children:

> 

Reviewer: Please remove that line.

Author: Removed.

Heads-up: openHAB's `MarkdownCheck` then flags line 3 with "The line before a Markdown list must be empty". It does not understand YAML frontmatter, so it is a false positive, and the hue README has the identical shape on main. The build still reports 0 errors.

:::tip Note
The metrics pipeline uses Micrometer's naming conventions (snake_case with `.` separators), not the OTel semantic conventions for metrics.
Use `CUMULATIVE` (the default) for most backends and when routing through an OTel Collector.
Use `DELTA` when pushing directly to a backend whose data model requires delta-encoded metrics — consult your backend's documentation.
:::

## Exported Traces (Event-Bus Spans)

The service subscribes to the entire openHAB event bus and emits one span per event, providing a complete activity timeline of your openHAB instance: item state changes, thing status transitions, rule executions, channel link events, and more.

Each span carries the following attributes:

| Attribute | Description |
|:---|:---|
| `event.type` | The openHAB event class name (e.g. `ItemStateChangedEvent`) |
| `event.topic` | The event bus topic (e.g. `openhab/items/MyLight/statechanged`) |
| `event.source` | The event source identifier |

Use `tracesSamplingRatio` to limit the exported volume on busy instances (e.g. `0.1` to export 10% of events).

## Deployment

The add-on supports two deployment patterns:

> **Direct to backend** — set `otlpURL` to your observability backend's OTLP ingest URL and `otlpHeaders` to the required authentication header.

Reviewer: ```suggestion
- **Direct to backend**: Set `otlpURL` to your observability backend's OTLP ingest URL and `otlpHeaders` to the required authentication header.
```

> Simple to set up; backend credentials are stored in openHAB's configuration.

Reviewer: ```suggestion
  Simple to set up; backend credentials are stored in openHAB's configuration.
```

> 

Reviewer: ```suggestion
```

> **Via an OTel Collector** — set `otlpURL` to the collector's HTTP endpoint (e.g. `http://localhost:4318`) and leave `otlpHeaders` empty.

Reviewer: ```suggestion
- **Via an OTel Collector**: Set `otlpURL` to the collector's HTTP endpoint (e.g. `http://localhost:4318`) and leave `otlpHeaders` empty.
```

> The collector receives all three signals from openHAB and forwards them to one or more backends.

Reviewer: ```suggestion
  The collector receives all three signals from openHAB and forwards them to one or more backends.
```

> This keeps backend credentials out of openHAB, allows fan-out to multiple backends, and handles metric temporality conversion.

Reviewer: ```suggestion
  This keeps backend credentials out of openHAB, allows fan-out to multiple backends, and handles metric temporality conversion.
```

> See the [OTel Collector example](doc/otel-collector.md) for a ready-to-use configuration.

Reviewer: ```suggestion
  See the [OTel Collector example](doc/otel-collector.md) for a ready-to-use configuration.
```

## Configuration

The OpenTelemetry service can be configured via Main UI (_Settings_ → _Add-on Settings_ → _OpenTelemetry Service_) or by using a configuration file (see [below](#configuration-file-example)).

### Configuration Parameters

#### Connection

| Parameter | Description | Default |
|:---|:---|:---|
| `otlpURL` | OTLP endpoint to push telemetry to. Set to a local OTel Collector (e.g. `http://localhost:4318`) or directly to a backend ingest URL. All per-signal endpoints are resolved against this base URL. | `http://localhost:4318` |
| `otlpHeaders` | Comma-separated authentication headers, e.g. `Authorization=Bearer token`. Only needed for direct-to-backend deployments — leave empty when using a collector. Stored as a masked secret. | |

:::tip Note

> If `otlpURL` uses `http://`, this is logged at startup. Use HTTPS in production to protect credentials in transit.

Reviewer: Most people won't have different environments, just one prod environment ...

```suggestion
If `otlpURL` uses `http://`, this is logged at startup. Use HTTPS to protect credentials in transit.
```

#### Traces

| Parameter | Description | Default |
|:---|:---|:---|
| `tracesEnabled` | Enable exporting event-bus spans to the OTLP endpoint | `false` |
| `tracesEndpoint` | Endpoint path, resolved against `otlpURL` | `/v1/traces` |
| `tracesSamplingRatio` | Fraction of event-bus spans to export (0.0 = none, 1.0 = all) | `1.0` |

### Configuration File Example

To configure the service via file, create or modify `$OPENHAB_CONF/services/opentelemetry.cfg`:

```ini
# Base URL of your OTLP endpoint or collector
otlpURL=http://localhost:4318

# Optional authentication headers (comma-separated key=value pairs)
# otlpHeaders=Authorization=Bearer mySecretToken

# --- Logs ---
logsEnabled=true
logsEndpoint=/v1/logs

# --- Metrics ---
metricsEnabled=true
metricsEndpoint=/v1/metrics
metricsInterval=PT60S
# CUMULATIVE works for most backends. Use DELTA if your backend requires delta-encoded metrics.
metricsAggregationTemporality=CUMULATIVE

# --- Traces (event-bus spans) ---
tracesEnabled=true
tracesEndpoint=/v1/traces
tracesSamplingRatio=1.0
```

> ## Scope and Limitations

Reviewer: This section ready a lot like AI-generated text. Can you please rephrase it manually and make it more human-like?

Author: Rewritten by hand. 

I gave the "Coexistence with the OTel Java Agent" section below it the same treatment, since it read the same way.

## Coexistence with the OTel Java Agent

The [OTel Java agent](https://opentelemetry.io/docs/zero-code/java/agent/) can complement this bundle by providing automatic HTTP server instrumentation (which adds `server`-kind root spans and enables RED metrics in observability backends), JDBC tracing, and other library-level coverage that the bundle cannot provide.

When the agent is detected at startup, this bundle routes its event-bus spans through the agent's already-registered `GlobalOpenTelemetry` pipeline rather than starting its own tracer provider, so both sources land in the same service entity in the backend.

To run the agent alongside openHAB, add to `/etc/default/openhab` (or the equivalent for your installation):

```bash
EXTRA_JAVA_OPTS="-javaagent:/path/to/opentelemetry-javaagent.jar \
  -Dotel.service.name=openHAB \
  -Dotel.exporter.otlp.protocol=http/protobuf \
  -Dotel.exporter.otlp.endpoint=http://localhost:4318 \
  -Dotel.metrics.exporter=none"
```

> Set `-Dotel.exporter.otlp.protocol` explicitly and make sure it matches your endpoint's port — `http/protobuf` typically listens on `4318`, `grpc` on `4317`. A protocol/port mismatch fails silently at the transport layer with no data arriving and no obvious error.

Reviewer: Put each sentence on a separate line.

```suggestion
Set `-Dotel.exporter.otlp.protocol` explicitly and make sure it matches your endpoint's port — `http/protobuf` typically listens on `4318`, `grpc` on `4317`.
A protocol/port mismatch fails silently at the transport layer with no data arriving and no obvious error.
```

#### Traces

| Parameter | Description | Default |
|:---|:---|:---|
| `tracesEnabled` | Enable exporting event-bus spans to the OTLP endpoint | `false` |
| `tracesEndpoint` | Endpoint path, resolved against `otlpURL` | `/v1/traces` |
| `tracesSamplingRatio` | Fraction of event-bus spans to export (0.0 = none, 1.0 = all) | `1.0` |

### Configuration File Example

To configure the service via file, create or modify `$OPENHAB_CONF/services/opentelemetry.cfg`:

```ini
# Base URL of your OTLP endpoint or collector
otlpURL=http://localhost:4318

# Optional authentication headers (comma-separated key=value pairs)
# otlpHeaders=Authorization=Bearer mySecretToken

# --- Logs ---
logsEnabled=true
logsEndpoint=/v1/logs

# --- Metrics ---
metricsEnabled=true
metricsEndpoint=/v1/metrics
metricsInterval=PT60S
# CUMULATIVE works for most backends. Use DELTA if your backend requires delta-encoded metrics.
metricsAggregationTemporality=CUMULATIVE

# --- Traces (event-bus spans) ---
tracesEnabled=true
tracesEndpoint=/v1/traces
tracesSamplingRatio=1.0
```

## Scope and Limitations

> The bundle reads telemetry from three places openHAB already offers: the OSGi log service, the Micrometer registry and the event bus.

Reviewer: ```suggestion
The add-on reads telemetry from three places openHAB already offers: the OSGi log service, the Micrometer metrics registry and the event bus.
```

No JVM agent is needed, but that also means it only sees what those three sources expose:

- Only openHAB's own logs, meters and events are exported.
  Calls a binding makes to a device or a REST API, persistence writes and MQTT traffic are not traced.
- Event-bus spans are flat.
  Each event gets its own root span, so you get a timeline of activity rather than call trees.
  No `traceparent` is passed between threads or over the network.

> - Nothing is captured before the bundle starts.

Reviewer: ```suggestion
- Nothing is captured before the add-on starts.
```
